### PR TITLE
Permissions: return permissions in request event calls

### PR DESCRIPTION
**Pull request by Pineirin**
Display edit/delete dropdown only for authorized users.
closes: inveniosoftware/invenio-requests#156

View of the user admin@test.ch
Screenshot from 2022-03-29 15-56-23

View of the user pablo@gmail.com
Screenshot from 2022-03-29 15-56-30

### CODE REVIEW

File: invenio_requests/services/schemas.py [hunk 1]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
...
         """Return key of type_schemas to use given object to dump."""
         return obj.type
+
+
+class Permissions():
+    """Add permissions to the dump data."""
+
+    permissions = fields.Method("get_permissions")
+
+    def is_self(self, obj):
```
Comment (kpsherva) [inline invenio_requests/services/schemas.py:130]: I think this name can be interpreted in too many ways, could you change it into something more explanatory, like `is_comment_creator` `in_event_creator` ?

File: setup.cfg [hunk 2]
```diff
@@ -35,6 +35,10 @@ tests =
     invenio-db[postgresql,mysql,versioning]>=1.0.13,<2.0
     pytest-invenio>=1.4.3
     sphinx>=4.2.0,<5
+    Flask<=2.0.3
```
Comment (kpsherva) [inline setup.cfg:38]: why did you need flask to be installed in the module ?

File: tests/services/events/test_request_events_service.py [hunk 3]
```diff
@@ -55,6 +55,8 @@ def test_simple_flow(
     # Update it
     data = read_item.to_dict()  # should be equivalent to events_service_data
     data["payload"]["content"] = "An edited comment"
+    # Remove permissions from data as they wont be persisted.
+    data.pop("permissions")
```
Comment (kpsherva) [inline tests/services/events/test_request_events_service.py:59]: not sure why tou pop the permissions?

File: invenio_requests/services/schemas.py [hunk 4]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
     def get_obj_type(self, obj):
         """Return key of type_schemas to use given object to dump."""
         return obj.type
+
+
+class Permissions():
```
Comment (kpsherva) [inline invenio_requests/services/schemas.py:125]: I think the name should follow the pattern of other schema classes fe `EventPermissionShema`

File: invenio_requests/assets/semantic-ui/js/invenio_requests/timelineEvent/TimelineEvent.js [hunk 5]
```diff
@@ -68,15 +71,19 @@ class TimelineEvent extends Component {
...
-                        </Dropdown.Item>
-                        <Dropdown.Item onClick={() => deleteComment()}>
-                          {i18next.t("Delete")}
-                        </Dropdown.Item>
+                        {canUpdate && (
+                          <Dropdown.Item onClick={() => toggleEditMode()}>
+                            {i18next.t("Edit")}
+                          </Dropdown.Item>
+                        )}
+                        {canUpdate && (
```
Comment (kpsherva) [inline invenio_requests/assets/semantic-ui/js/invenio_requests/timelineEvent/TimelineEvent.js:82]:
```suggestion
                        {canDelete && (
```

**Review by kpsherva — COMMENTED** [review]
just because the solution is changing the expected payload of the event, was this discussed with anyone, or taken into the account in the RFCs?

File: invenio_requests/services/schemas.py [hunk 6]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
...
         return obj.type
+
+
+class EventPermissionShema():
+    """Add permissions to the dump data."""
+
+    permissions = fields.Method("get_permissions")
+
+    def in_event_creator(self, obj):
+        """True if the creator of the record is the identified user."""
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:131]:
```suggestion
        """True if the creator of the record is the current_user."""
```

File: invenio_requests/services/schemas.py [hunk 7]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
...
         """Return key of type_schemas to use given object to dump."""
         return obj.type
+
+
+class EventPermissionShema():
+    """Add permissions to the dump data."""
+
+    permissions = fields.Method("get_permissions")
+
+    def in_event_creator(self, obj):
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:130]:
```suggestion
    def is_event_creator(self, obj):
```

File: invenio_requests/services/schemas.py [hunk 8]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
...
+        """True if the creator of the record is the identified user."""
+        if "is_self" not in self.context:
+            current_identity = self.context["identity"]
+            user_id = obj.created_by._resolve().get_id()
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:134]: I am not sure about this code, why the `_resolve()`?
What about:
```
def is_event_creator(self, obj):
    """...."""
    current_identity = self.context["identity"]
    user_id = obj.created_by.get('user', '')
    return current_identity.id is not None and str(user_id) == str(current_identity.id)
```

File: invenio_requests/services/schemas.py [hunk 9]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
...
+            self.context["is_self"] = (
+                user_id is not None
+                and current_identity.id is not None
+                and str(user_id) == str(current_identity.id)
+            )
+        return self.context["is_self"]
+
+    def get_permissions(self, obj):
+        """Get certain permissions of the user with the current hit."""
+        return {
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:144]:
```suggestion
        is_creator = self.in_event_creator(obj)
        return {
            can_update_comment: is_creator,
            can_delete_commet: is_creator,
```

File: invenio_requests/services/events/service.py [hunk 10]
```diff
@@ -11,16 +11,22 @@
...
 
 from ...records.api import RequestEventType
+from ..schemas import RequestEventDumpSchema
 
 
 class RequestEventsService(RecordService):
     """Request Events service."""
 
+    @property
+    def member_dump_schema(self):
```
Comment (ntarocco) [inline invenio_requests/services/events/service.py:26]: rename?

File: invenio_requests/services/schemas.py [hunk 11]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
...
+
+class CommentDumpSchema(CommentSchema, EventPermissionShema):
+    """Comment dump schema."""
+
+
+class NoExtrasDumpSchema(NoExtrasSchema, EventPermissionShema):
+    """Comment dump schema."""
+
+
+class RequestEventDumpSchema(BaseRecordSchema, OneOfSchema):
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:158]: I am not sure this duplication of schemas is needed, let's discuss IRL.

File: invenio_requests/services/schemas.py [hunk 12]
```diff
@@ -120,3 +120,54 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
     def get_obj_type(self, obj):
         """Return key of type_schemas to use given object to dump."""
         return obj.type
+
+
+class EventPermissionShema():
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:125]: Implementation is nice, good work, I am just thinking that `can_update_comment` or `can_delete_comment` are permissions that are configurable, while here you are hardcoding that the current logged in user must be the creator.

**Review by jrcastro2 — APPROVED** [review]
LGTM! :rocket:

File: tests/services/events/test_request_events_service.py (hunk 3 above)
Comment (lnielsen) [reply tests/services/events/test_request_events_service.py:59]: Is it because you cannot send ``permissions``  key back? If, so then it's likely a bug that should be fixed in the schema.......

File: invenio_requests/services/schemas.py [hunk 13]
```diff
@@ -120,3 +123,47 @@ class RequestEventSchema(BaseRecordSchema, OneOfSchema):
     def get_obj_type(self, obj):
         """Return key of type_schemas to use given object to dump."""
         return obj.type
+
+
+class EventPermissionShema(Schema):
```
Comment (ntarocco) [inline invenio_requests/services/schemas.py:128]: It looks to me very Comment specific, probably rename to:
```suggestion
class CommentPermissionShema(Schema):
```

**Review by ntarocco — APPROVED** [review]
